src/cargo/core/interning.rs

```diff
@@ -24,65 +24,71 @@ lazy_static! {
         RwLock::new(HashSet::new());
 }
 
-#[derive(Eq, PartialEq, Clone, Copy)]
+#[derive(Clone, Copy)]
 pub struct InternedString {
-    ptr: *const u8,
-    len: usize,
+    inner: &'static str,
 }
 
+impl PartialEq for InternedString {
+    fn eq(&self, other: &InternedString) -> bool {
+        fn identity(s: &InternedString) -> (*const u8, usize) {
+            (s.inner.as_ptr(), s.inner.len())
+        }
+        identity(self) == identity(other)
```

Comment:
`ptr::eq(self.inner, other.inner)`?

Reply:
Wow, I haven't realized that `*const str` is a thing!

Reply:
@matklad I am willing to approve when you are happy. Do you want to make this change, or should I say the magick spell?